### PR TITLE
Adjuster fixes and added GUI prompt for applying last settings

### PR DESCRIPTION
**Pull request by Robinde67**
Adding to the sprite pool seems to give some unexpected behavior, like skipping adjusting or deleting the settings from persistent storage, though the former issue seems to be in upstream as well.
Moved and rewrote code from get_adjuster_settings in Utils in an attempt to make it more scalable for other adjusters in the future (no more LttP specific code in get_adjuster_settings)

### CODE REVIEW

**Comment by Berserker66** [comment]
"Adding to the sprite pool seems to give some unexpected behavior, like skipping adjusting or deleting the settings from persistent storage, though the former issue seems to be in upstream as well."
What exactly do you mean by that?

**Comment by Robinde67** [comment]
When running upstream it just seems ignore the settings if sprite-pool is present, PR does the same and also delete the entire entry from persistent storage, as long as sprite-pool isn't set it works just fine throughout all my tests, but i'm not sure why it behaves the way it does when sprite-pool is present, though i'm also not sure what the setting itself does

**Comment by Berserker66** [comment]
sprite pool is the set of sprites to pick from for "random sprite on event". If you found a bug there it'd be nice to also fix it.

**Comment by Robinde67** [comment]
The bug may be because i haven't chosen any events for them to trigger on then, i'll take a look at it when I've recovered

**Comment by Robinde67** [comment]
The error was caused by my Save settings function incorrectly saving the wrong variable making it unreadable, and that was the persistent storage yaml I used to test on upstream, it now works as intended.

File: SNIClient.py [hunk 1]
```diff
@@ -1087,14 +1088,125 @@ async def main():
...
         elif args.diff_file.endswith((".apbp", "apz3")):
-            adjustedromfile, adjusted = Utils.get_adjuster_settings(romfile, gui_enabled)
-            if adjusted:
-                try:
-                    shutil.move(adjustedromfile, romfile)
-                    adjustedromfile = romfile
-                except Exception as e:
-                    logging.exception(e)
-            asyncio.create_task(run_game(adjustedromfile if adjusted else romfile))
+            lastSettings = Utils.get_adjuster_settings("LttP")
```
Comment (Berserker66) [inline SNIClient.py:1091]: internal name for LttP is nowadays the full "A Link to the Past".

File: SNIClient.py [hunk 2]
```diff
@@ -1087,14 +1088,125 @@ async def main():
...
+                                sprite_pool[sprite] += 1
+                            else:
+                                sprite_pool[sprite] = 1
+                        if sprite_pool:
+                            printed_options["sprite_pool"] = sprite_pool
+                    import pprint
+
+                    if gui_enabled:
+            
+                        from tkinter import Tk, PhotoImage, Label, LabelFrame, Frame, Button
```
Comment (Berserker66) [inline SNIClient.py:1114]: this might be best to extract out as a function, it makes following the flow of the outer function somewhat difficult

**Review by Berserker66 — CHANGES_REQUESTED** [review]

File: SNIClient.py (hunk 1 above)
Comment (Robinde67) [reply SNIClient.py:1091]: I can change it to "A Link to the Past", that will not look good in persistent storage though
`last_settings_A Link to the Past: !!python/object:argparse.Namespace`
I could probably replace spaces with underscores in the function

**Comment by Berserker66** [comment]
Just a time update; this PR wasn't quite ready in time for the release candidates. This weekend I'll be busy with 0.2.3 release and the new async. So will look at this proper likely early next week. Sorry for the delay on my part.

**Comment by Robinde67** [comment]
There is no rush, take the time you need.

**Comment by Berserker66** [comment]
Since the ask option won't work when windows compiled
image
Maybe check for if sys.stdin and only show the button if it exists?

File: SNIClient.py [hunk 3]
```diff
@@ -1131,6 +1129,126 @@ async def main():
...
+        if not hasattr(lastSettings, 'auto_apply') or 'ask' in lastSettings.auto_apply:
+
+            whitelist = {"music", "menuspeed", "heartbeep", "heartcolor", "ow_palettes", "quickswap",
+                        "uw_palettes", "sprite", "sword_palettes", "shield_palettes", "hud_palettes",
+                        "reduceflashing", "deathlink"}
+            printed_options = {name: value for name, value in vars(lastSettings).items() if name in whitelist}
+            if hasattr(lastSettings, "sprite_pool"):
+                sprite_pool = {}
+                for sprite in getattr(lastSettings, "sprite_pool"):
+                    if sprite in sprite_pool:
```
Comment (Berserker66) [inline SNIClient.py:1147]: I think you might want collection.Counter(last_settings.sprite_pool), which is a subclass of a dict

File: SNIClient.py [hunk 4]
```diff
@@ -1131,6 +1129,126 @@ async def main():
...
+                    if sprite_pool:
+                        printed_options["sprite_pool"] = sprite_pool
+            import pprint
+
+            if gui_enabled:
+            
+                from tkinter import Tk, PhotoImage, Label, LabelFrame, Frame, Button
+                applyPromptWindow = Tk()
+                applyPromptWindow.resizable(False, False)
+                applyPromptWindow.protocol('WM_DELETE_WINDOW',lambda: onButtonClick())
```
Comment (Berserker66) [inline SNIClient.py:1160]: without testing admittedly, but I have a feeling `applyPromptWindow.protocol('WM_DELETE_WINDOW', onButtonClick)` should work here

**Review by Berserker66 — CHANGES_REQUESTED** [review]

File: SNIClient.py [hunk 5]
```diff
@@ -1087,14 +1088,11 @@ async def main():
...
-            adjustedromfile, adjusted = Utils.get_adjuster_settings(romfile, gui_enabled)
-            if adjusted:
-                try:
-                    shutil.move(adjustedromfile, romfile)
-                    adjustedromfile = romfile
-                except Exception as e:
-                    logging.exception(e)
-            asyncio.create_task(run_game(adjustedromfile if adjusted else romfile))
+            adjustedromfile, adjusted = get_alttp_settings(romfile)
+            if(adjusted):
```
[the remark this answers is not in the document]
Comment (Robinde67) [reply SNIClient.py:1092]: Leftover code after the function breakout.

File: SNIClient.py [hunk 6]
```diff
@@ -1131,6 +1129,126 @@ async def main():
...
+        choice = 'no'
+        if not hasattr(lastSettings, 'auto_apply') or 'ask' in lastSettings.auto_apply:
+
+            whitelist = {"music", "menuspeed", "heartbeep", "heartcolor", "ow_palettes", "quickswap",
+                        "uw_palettes", "sprite", "sword_palettes", "shield_palettes", "hud_palettes",
+                        "reduceflashing", "deathlink"}
+            printed_options = {name: value for name, value in vars(lastSettings).items() if name in whitelist}
+            if hasattr(lastSettings, "sprite_pool"):
+                sprite_pool = {}
+                for sprite in getattr(lastSettings, "sprite_pool"):
```
[the remark this answers is not in the document]
Comment (Robinde67) [reply SNIClient.py:1146]: https://github.com/ArchipelagoMW/Archipelago/blob/a6cca3094df3780891e52505058048a9a2e4cb42/Utils.py#L321
Not my code but I'll fix it.

File: SNIClient.py (hunk 3 above)
Comment (Robinde67) [reply SNIClient.py:1147]: Same as previous, not my code, not sure what the original intention of that code is.

File: SNIClient.py (hunk 4 above)
Comment (Robinde67) [reply SNIClient.py:1160]: Been a few weeks but I do recall not using lambda at first and it did not work as intended, don't remember if the window just closed immediately or if I got a function not defined, but I had to use lambda

**Review by Berserker66 — APPROVED** [review]

**Comment by Berserker66** [comment]
Thank you!

**Pull request merged**